### PR TITLE
[repl] Add commissioning with QR code / manual setup payload

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem

`devCtrl->PairDevice` with SetupPayload is not exposed

#### Change overview

Expose it

#### Testing

- Tested manually in WiFi only & BLE only enviroments, this is a trivial API exporting PR.

### CODE REVIEW

**Review by Damian-Nordic — APPROVED** [review]

**Review by yunhanw-google — COMMENTED** [review]
Can we add automatic test for this?

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 1]
```diff
@@ -290,6 +290,29 @@ def CheckTestCommissionerCallbacks(self):
     def CheckTestCommissionerPaseConnection(self, nodeid):
         return self._dmLib.pychip_TestPaseConnection(nodeid)
 
+    def CommissionWithCode(self, setupPayload: str, nodeid: int):
```
Comment (yunhanw-google) [inline src/controller/python/chip/ChipDeviceCtrl.py:293]: maybe we can start from here to remove ChipDeviceCtrl.py and start to create the Commision.py?

Comment (erjiaqing) [reply src/controller/python/chip/ChipDeviceCtrl.py:293]: Based on the use of devCtrl class, I would suggest a separate refactor to remove the legacy ZCL code first.

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 2]
```diff
@@ -290,6 +290,29 @@ def CheckTestCommissionerCallbacks(self):
...
+        # for the commissioning complete event, not just any callback.
+        self.state = DCState.RENDEZVOUS_ONGOING
+
+        self._ChipStack.commissioningCompleteEvent.clear()
+
+        self._ChipStack.CallAsync(
+            lambda: self._dmLib.pychip_DeviceController_ConnectWithCode(
+                self.devCtrl, setupPayload, nodeid)
+        )
+        # Wait up to 5 additional seconds for the commissioning complete event
```
Comment (tcarmelveilleux) [inline src/controller/python/chip/ChipDeviceCtrl.py:308]: This is not long enough. Why 5 seconds? Why not base it on the sleepy interval, etc ?

File: src/controller/python/test/test_scripts/commissioning_test.py [hunk 3]
```diff
@@ -88,20 +115,28 @@ def main():
...
     FailIfNot(test.SetNetworkCommissioningParameters(dataset=TEST_THREAD_NETWORK_DATASET_TLV),
               "Failed to finish network commissioning")
 
-    logger.info("Testing key exchange")
-    FailIfNot(test.TestKeyExchange(ip=options.deviceAddress,
-                                   setuppin=20202021,
-                                   nodeid=1),
-              "Failed to finish key exchange")
+    if options.deviceAddress:
+        logger.info("Testing key exchange (IP)")
```
Comment (tcarmelveilleux) [inline src/controller/python/test/test_scripts/commissioning_test.py:125]: Why "key exchange" ? It's not key exchange, it's full commissioning

**Review by tcarmelveilleux — APPROVED** [review]

File: src/controller/python/test/test_scripts/commissioning_test.py (hunk 3 above)
Comment (erjiaqing) [reply src/controller/python/test/test_scripts/commissioning_test.py]: Renamed key exchange -> commissioning

File: src/controller/python/test/test_scripts/commissioning_test.py [hunk 4]
```diff
@@ -68,6 +68,33 @@ def main():
...
+        help="Setup Payload (manual pairing code or QR code content)",
+        metavar="<setup-payload>"
+    )
+    optParser.add_option(
+        "--nodeid",
+        action="store",
+        dest="nodeid",
+        default=1,
+        type=int,
+        help="Node ID of the device",
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply src/controller/python/test/test_scripts/commissioning_test.py]: Yes, this is the node id that will be issued to the device. Updated.

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
@erjiaqing - cirque and repl tests seemed failed. I restarted them just in case, however I believe that this PR changes python and both python unit tests fail, so likely a real failure.

**Comment by erjiaqing** [comment]
> @erjiaqing - cirque and repl tests seemed failed. I restarted them just in case, however I believe that this PR changes python and both python unit tests fail, so likely a real failure.

Yes, this is a real failure, I think we need a huge refactor on commissioning code since it is a mixture of old "rendezvous" and current pairer implementation. We need to remove the logic from the old rendezvous code.

**Pull request merged**